impl/src/fmt.rs

```diff
@@ -94,6 +101,122 @@ impl Display<'_> {
         self.args = args;
         self.has_bonus_display = has_bonus_display;
     }
+
+    pub fn iter_fmt_types(&'a self, fields: &'a [Field]) -> Vec<DisplayFormatMarking<'a>> {
```

Comment:
It's problematic that this 172 lines of changes in fmt.rs is duplicating the entire format string parser in a way that is completely detached from (and incompatible with) the existing format string processing.

Take a look at my version which sticks a comparatively tiny piece of code (24 lines) in the existing format string codepath to wire up implied bounds as a natural part of processing the format string.

Reply:
Yeah, using a member on the Field output was a better move. I like that it also allows bounds for the other display types. I think the one thing I'd change given infinite time would be splitting `Display<'_>` into pre-expanded and post-expanded types, and keeping the template and non-template sections around as an ADT, then resolving the field rename hacks on demand.